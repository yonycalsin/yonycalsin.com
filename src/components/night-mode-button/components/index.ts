export * from './moon-icon'
export * from './sun-icon'
