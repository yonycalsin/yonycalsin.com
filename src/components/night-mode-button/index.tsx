import * as React from 'react'
import clsx from 'clsx'

import s from './night-mode-button.module.css'

const STORAGE_KEY = 'color-mode'

const NightModeButton = () => {
  function onToggleDark() {
    const isDark = document.documentElement.classList.contains('dark')

    localStorage.setItem(STORAGE_KEY, isDark ? 'light' : 'dark')

    if (isDark) {
      document.documentElement.classList.remove('dark')
    } else {
      document.documentElement.classList.add('dark')
    }
  }

  return (
    <button
      className="appearance-none outline-none bg-warning-200 dark:bg-gray-700 rounded-full h-20 w-20 fixed overflow-hidden -left-10 -top-10 border border-transparent hover:border-primary z-10"
      onClick={onToggleDark}
      style={{
        outline: 'none',
      }}
    >
      <div
        className="top-1 bottom-1 flex flex-col items-stretch justify-between absolute w-8 transform rotate-135 dark:-rotate-45 transition-transform duration-500"
        style={{
          left: 'calc(50% - 2rem)',
        }}
      >
        <span className={clsx(s.item, 'w-7 h-7')}></span>
        <span
          className={clsx(s.item, 'w-7 h-7')}
          style={{
            backgroundPosition: 'bottom',
          }}
        ></span>
      </div>
    </button>
  )
}

export default NightModeButton
