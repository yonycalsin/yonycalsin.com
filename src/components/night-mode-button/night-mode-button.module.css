.item {
  background-image: url(../../assets/images/sun-moon.png);
  background-size: theme('spacing.7');
  transform: rotate(45deg);
}
