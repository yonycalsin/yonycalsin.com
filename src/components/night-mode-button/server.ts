'use server'

import { CookieStorage } from 'constants/storage'
import { cookies } from 'next/headers'

// eslint-disable-next-line @typescript-eslint/require-await
export async function toggle() {
  const mode = cookies().get(CookieStorage.DARK)

  if (mode) {
    cookies().delete(CookieStorage.DARK)
  } else {
    cookies().set(CookieStorage.DARK, 'true')
  }
}
