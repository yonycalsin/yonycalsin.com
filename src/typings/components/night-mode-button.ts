interface SunIconProps {
  className: string
}

interface MoonIconProps {
  className: string
}

export type { MoonIconProps, SunIconProps }
