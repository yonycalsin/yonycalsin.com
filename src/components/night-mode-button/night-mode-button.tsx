import * as React from 'react'
import { Box, useColorMode } from '@chakra-ui/react'

import { analyticsEvents } from '~/analytics/events'
import { analytics } from '~/analytics/google-analytics'

import { MoonIcon } from './components/moon-icon'
import { SunIcon } from './components/sun-icon'

export const NightModeButton = () => {
  const { colorMode, toggleColorMode } = useColorMode()

  return (
    <>
      <Box
        role="button"
        as="button"
        position="fixed"
        left={{
          base: 4,
          md: 8,
        }}
        top={{
          base: 20,
          md: 8,
        }}
        fontSize={{
          base: '2xl',
          md: '4xl',
        }}
        zIndex="sticky"
        onClick={() => {
          analytics.event({
            action: analyticsEvents.TOGGLE_NIGHT_MODE_BUTTON,
            category: 'ui-theme',
            label: colorMode,
          })

          toggleColorMode()
        }}
      >
        {colorMode === 'dark' ? <SunIcon className="transition-all" /> : <MoonIcon className="transition-all" />}
      </Box>
    </>
  )
}
