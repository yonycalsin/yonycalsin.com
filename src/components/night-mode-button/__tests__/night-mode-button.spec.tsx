import type { useColorMode } from '@chakra-ui/react'
import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import { NightModeButton } from '../night-mode-button'

const mockToggleColorMode = jest.fn()

const mockUseColorMode = jest.fn() as jest.Mock<Partial<ReturnType<typeof useColorMode>>>

jest.mock('@chakra-ui/react', () => ({
  ...jest.requireActual<typeof import('@chakra-ui/react')>('@chakra-ui/react'),
  useColorMode: () => mockUseColorMode(),
}))

const setup = () => {
  return render(<NightModeButton />)
}

describe('Components / NightModeButton', () => {
  afterEach(() => {
    jest.clearAllMocks()
  })

  it('renders the moon icon', () => {
    mockUseColorMode.mockReturnValue({
      colorMode: 'dark',
    })

    const view = setup()

    expect(view.baseElement).toMatchSnapshot()
  })

  it('renders the sun icon', () => {
    mockUseColorMode.mockReturnValue({
      colorMode: 'light',
    })

    const view = setup()

    expect(view.baseElement).toMatchSnapshot()
  })

  it('clicks on the toggle button', async () => {
    mockUseColorMode.mockReturnValue({
      colorMode: 'dark',
      toggleColorMode: mockToggleColorMode,
    })

    setup()

    const button = screen.getByRole('button')

    expect(button).toBeInTheDocument()

    await userEvent.click(button)

    expect(mockToggleColorMode).toHaveBeenCalledTimes(1)
  })
})
