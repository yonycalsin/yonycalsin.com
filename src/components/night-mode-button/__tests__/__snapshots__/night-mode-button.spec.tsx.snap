// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components / NightModeButton renders the moon icon 1`] = `
<body>
  <div>
    <button
      class="emotion-0"
      role="button"
    >
      <svg
        class="transition-all"
        fill="none"
        height="1em"
        style="transform-origin: 50% 50%;"
        transform-origin="50% 50%"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx="11.9998"
          cy="11.9998"
          fill="currentColor"
          r="5.75375"
          style="transform: scale(1.5); transform-origin: 0px 0px;"
          transform-origin="0px 0px"
        />
        <g
          style="transform: rotate(45deg); transform-origin: 0px 0px;"
          transform-origin="0px 0px"
        >
          <circle
            cx="3.08982"
            cy="6.85502"
            fill="currentColor"
            r="1.71143"
            transform="rotate(-60 3.08982 6.85502)"
          />
          <circle
            cx="3.0903"
            cy="17.1436"
            fill="currentColor"
            r="1.71143"
            transform="rotate(-120 3.0903 17.1436)"
          />
          <circle
            cx="12"
            cy="22.2881"
            fill="currentColor"
            r="1.71143"
          />
          <circle
            cx="20.9101"
            cy="17.1436"
            fill="currentColor"
            r="1.71143"
            transform="rotate(-60 20.9101 17.1436)"
          />
          <circle
            cx="20.9101"
            cy="6.8555"
            fill="currentColor"
            r="1.71143"
            transform="rotate(-120 20.9101 6.8555)"
          />
          <circle
            cx="12"
            cy="1.71143"
            fill="currentColor"
            r="1.71143"
          />
        </g>
      </svg>
    </button>
  </div>
</body>
`;

exports[`Components / NightModeButton renders the sun icon 1`] = `
<body>
  <div>
    <button
      class="emotion-0"
      role="button"
    >
      <svg
        class="transition-all"
        height="1em"
        viewBox="0 0 50 50"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M 43.81 29.354 C 43.688 28.958 43.413 28.626 43.046 28.432 C 42.679 28.238 42.251 28.198 41.854 28.321 C 36.161 29.886 30.067 28.272 25.894 24.096 C 21.722 19.92 20.113 13.824 21.683 8.133 C 21.848 7.582 21.697 6.985 21.29 6.578 C 20.884 6.172 20.287 6.022 19.736 6.187 C 10.659 8.728 4.691 17.389 5.55 26.776 C 6.408 36.163 13.847 43.598 23.235 44.451 C 32.622 45.304 41.28 39.332 43.816 30.253 C 43.902 29.96 43.9 29.647 43.81 29.354 Z"
          fill="currentColor"
          style="transform: scale(0.6) rotate(90deg); transform-origin: 0px 0px;"
          transform-origin="0px 0px"
        />
      </svg>
    </button>
  </div>
</body>
`;
